### PR TITLE
DO NOT INTEGRATE.

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
Draft pull request.

### CODE REVIEW

File: src/python/grpcio/grpc/_channel.py [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+    with state.condition:
+      if state.code is None:
+        operations = (cygrpc.operation_send_close_from_client(),)
+        call.start_batch(cygrpc.Operations(operations), event_handler)
+        state.due.add(cygrpc.OperationType.send_close_from_client)
+  thread = threading.Thread(target=consume_request_iterator)
+  thread.start()
+
+
+class _Rendezvous(grpc.Future, grpc.Call, grpc.RpcError):
```
Comment (soltanmm) [inline src/python/grpcio/grpc/_channel.py]: I still find this whole inheritance-instead-of-composition thing enormously confusing, but without statically typed attributes my personal reservations are kinda moot...

File: src/python/grpcio/tests/unit/_rpc_test.py [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+    multi_callable = _stream_unary_multi_callable(self._channel)
+    with self._control.fail():
+      with self.assertRaises(grpc.RpcError) as exception_context:
+        multi_callable(
+            request_iterator,
+            metadata=((b'test', b'FailedStreamRequestBlockingUnaryResponse'),))
+
+    self.assertIs(grpc.StatusCode.UNKNOWN, exception_context.exception.code())
+
+  @unittest.skip('memory leaks!')
```
Comment (soltanmm) [inline src/python/grpcio/tests/unit/_rpc_test.py]: This was the memory leak that you mentioned catching in https://github.com/grpc/grpc/pull/6327#issuecomment-220799424 ?

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_rpc_test.py]: It was. Are you reviewing these four commits commit-by-commit? You shouldn't be seeing any skips at head-of-line-of-development.

File: src/python/grpcio/grpc/_channel.py (hunk 1 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_channel.py]: It threads a very narrow needle eye with the way it has one `cancel` method satisfying two different `cancel` specifications from two different interfaces with only coincidentally compatible semantics.

I think I'd like to tease this class apart in the future.

File: src/python/grpcio/grpc/_channel.py [hunk 3]
```diff
@@ -194,25 +189,23 @@ def consume_request_iterator():
...
-        operations = (cygrpc.operation_send_close_from_client(),)
+        operations = (cygrpc.operation_send_close_from_client(_EMPTY_FLAGS),)
         call.start_batch(cygrpc.Operations(operations), event_handler)
         state.due.add(cygrpc.OperationType.send_close_from_client)
   thread = threading.Thread(target=consume_request_iterator)
   thread.start()
 
 
-class _Rendezvous(grpc.Future, grpc.Call, grpc.RpcError):
+class _Rendezvous(grpc.RpcError, grpc.Future, grpc.Call):
```
Comment (soltanmm) [inline src/python/grpcio/grpc/_channel.py]: Got bitten by MRO and `grpc.RpcError` inheriting from `Exception`?

File: src/python/grpcio/grpc/__init__.py [hunk 4]
```diff
@@ -734,6 +803,172 @@ def stop(self, grace):
...
+      be omitted or be False.
+    require_client_auth: A boolean indicating whether or not to require clients
+      to be authenticated. May only be True if root_certificates is not None.
+
+  Returns:
+    A ServerCredentials for use with an SSL-enabled Server.
+  """
+  if len(private_key_certificate_chain_pairs) == 0:
+    raise ValueError(
+        'At least one private key-certificate chain pairis required!')
```
Comment (soltanmm) [inline src/python/grpcio/grpc/__init__.py]: "Typo in Paris" could totally be an electronic music album title.

File: src/python/grpcio/grpc/_channel.py (hunk 3 above)
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_channel.py]: `pylint`, actually - I'm pretty sure the interpreter was just fine with the order as it was.

**Comment by jtattermusch** [comment]
Fine to close?

**Comment by jtattermusch** [comment]
Closing for inactivity, reopen if still needed.

**Pull request closed without merging**